feat: generate endpoints using replica numbers (#45862)
sourcegraph/sourcegraph: CHANGELOG.md

- Added a button "Reindex now" to the index status page. Admins can now force an immediate reindex of a repository. [#45533](https://github.com/sourcegraph/sourcegraph/pull/45533)
- Added an option "Unlock user" to the actions dropdown on the Site Admin Users page. Admins can unlock user accounts that wer locked after too many sign-in attempts. [#45650](https://github.com/sourcegraph/sourcegraph/pull/45650)
- Templates for certain emails sent by Sourcegraph are now configurable via `email.templates` in site configuration. [#45671](https://github.com/sourcegraph/sourcegraph/pull/45671)

> - Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the correct environment variables (`GITSERVER_REPLICA_NUMBER`, `INDEXED_SEARCH_REPLICA_NUMBER`, `SEARCHER_REPLICA_NUMBER`, `SYMBOLS_REPLICA_NUMBER`) are provided in Kubernetes and Docker-Compose deployments

Reviewer: Rename these to `*_REPLICA_COUNT` instead of `*_REPLICA_NUMBER` to be more concise.

> - Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service is provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`) must be set to `kubernetes` or `docker-compose` to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)

Reviewer: Endpoints for Gitserver, Indexed-search, Searcher, and Symbols are generated automatically when the replica counts for each service are provided using the correct environment variables (GITSERVER_REPLICA_COUNT, INDEXED_SEARCH_REPLICA_COUNT, SEARCHER_REPLICA_COUNT, SYMBOLS_REPLICA_COUNT) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (SEARCHER_URL, SYMBOLS_URL, INDEXED_SEARCH_SERVERS, SRC_GIT_SERVERS) must be set to kubernetes or docker-compose to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)

> - Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service is provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`) must be set to `kubernetes` or `docker-compose` to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)

Reviewer: ```suggestion
Endpoints for Gitserver, Indexed-search, Searcher, and Symbols are generated automatically when the replica counts for each service are provided using the correct environment variables (GITSERVER_REPLICA_COUNT, INDEXED_SEARCH_REPLICA_COUNT, SEARCHER_REPLICA_COUNT, SYMBOLS_REPLICA_COUNT) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (SEARCHER_URL, SYMBOLS_URL, INDEXED_SEARCH_SERVERS, SRC_GIT_SERVERS) must be set to kubernetes or docker-compose to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)
```

> - Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service is provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`) must be set to `kubernetes` or `docker-compose` to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)

Reviewer: ```suggestion
- Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service are provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`) must be set to `kubernetes` or `docker-compose` to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)
```

> - Endpoints for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` are generated automatically when the replica counts for each service is provided using the correct environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) in Docker-Compose and Kubernetes (when RBAC is disabled) deployments. The values for the URL environment variables for each service (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`) must be set to `kubernetes` or `docker-compose` to enable this feature. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)

Reviewer: I would suggest rephrasing to begin with mentioning the new environment variables, as this is the most relevant piece of information.

Added new environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) to the `sourcegraph-frontend` service to inform it of the replica count of sharded statefulsets. It uses these variables to determine the network address of each shard. These variables are intended to replace existing variables like `SRC_GIT_SERVERS`, which pass the shard addresses as a string, rather than relying on k8s naming conventions. If both `SRC_GIT_SERVERS` and `GITSERVER_REPLICA_COUNT` are set, `sourcegraph-frontend` will throw an error and fail to start.

Author: @beyang good point, thanks! What do you think about this?

```suggestion
- Added new environment variables (`GITSERVER_REPLICA_COUNT`, `INDEXED_SEARCH_REPLICA_COUNT`, `SEARCHER_REPLICA_COUNT`, `SYMBOLS_REPLICA_COUNT`) to the sourcegraph-frontend service to help it determine the network address of each shard in the sharded statefulsets based on replica counts. These new variables are intended to replace old ones like `SRC_GIT_SERVERS`. Sourcegraph will fail to start when both old and new variables are set. [#45862](https://github.com/sourcegraph/sourcegraph/pull/45862)
```

- Added a button "Reindex now" to the index status page. Admins can now force an immediate reindex of a repository. [#45533](https://github.com/sourcegraph/sourcegraph/pull/45533)
- Added an option "Unlock user" to the actions dropdown on the Site Admin Users page. Admins can unlock user accounts that wer locked after too many sign-in attempts. [#45650](https://github.com/sourcegraph/sourcegraph/pull/45650)

> - Templates for certain emails sent by Sourcegraph are now configurable via `email.templates` in site configuration. [#45671](https://github.com/sourcegraph/sourcegraph/pull/45671)

Reviewer 2: intentionally added?

Author: Nope! Removed! Thanks for catching that!
